### ISSUE TITLE
Out of range loads data values

### ISSUE THREAD

**Issue by TonyM1958** [alpriest/EnergyStats-Android#19]
Looking at out of range 'loads' values, here's what I'm currently doing to my data in HA. Top is fixed value, bottom is raw value from the inverter:

![image](https://github.com/alpriest/EnergyStats-Android/assets/63789168/211b6481-c459-4311-abfe-69be8c192329)

The out of range value appears to affect the top 3 hex digits of the 32-bit value (in my case at least). Here's a code snippet that shows the problem and possibly how you might fix it:

![image](https://github.com/alpriest/EnergyStats-Android/assets/63789168/dad3025a-0a52-4be8-b65f-57218730b579)

Basically - work out if the value has 0 in the top 3 hex digits, mask these off and subtract the result from the out of range number to leave the lower 5 hex digits and divide by 10 to restore the scale. The code goes around the houses to do this so you can see the intermediate results - but I'm sure you get the picture? You could just invert the mask and apply this to the 32-bit value after x 10 and then /10 again...

**Comment by TonyM1958** [comment]
Here's what the report data looks like if I don't fix the values:

![image](https://github.com/alpriest/EnergyStats-Android/assets/63789168/d64df13b-e189-4b09-9043-65a69fe7dd3f)

and with the fix:

![image](https://github.com/alpriest/EnergyStats-Android/assets/63789168/d0d46aa7-5d5b-4427-8dda-c95fa5de0e3d)

**Comment by TonyM1958** [comment]
And for month without the fix:

![image](https://github.com/alpriest/EnergyStats-Android/assets/63789168/f7c7d9bc-39fe-4e0c-b1f4-b9f4242a3f3e)

and year:

![image](https://github.com/alpriest/EnergyStats-Android/assets/63789168/0e65b9d7-f687-4349-8fc0-ee98be41bcf9)

**Comment by TonyM1958** [comment]
Just playing around with a few values posted to FB, suggests you might need to mask the top word of the 32-bit value to get the right value:

![image](https://github.com/alpriest/EnergyStats-Android/assets/63789168/0a9a6b85-e40f-4a98-ad0f-eb050c031e01)

As you say on FB, you could define a value to subtract or a mask to use with the underlying value. The problem with masking the complete top word is the you limit the value to 6,553,5kwh a year - which is probably starting to overlap with some peoples actual use.

**Comment by alpriest** [comment]
I've added this in the latest builds and also guarded against negative values which get inverted otherwise.

**Comment by TonyM1958** [comment]
You are probably on to this one already, but in the latest app -  negative values are being treated as unsigned, resulting in large values being shown when charging from grid.

For example, I'm seeing max Output energy 6548.4 for last night when the actual was around -5.2.

Don't know if this is related to guarding against -ve values as you mentioned above?

**Comment by alpriest** [comment]
Yup, broke in 1.86 I think, then should be resolved in 1.89.On 24 Oct 2023, at 19:02, TonyM1958 ***@***.***> wrote:﻿
You are probably on to this one already, but in the latest app -  negative values are being treated as unsigned, resulting in large values being shown when charging from grid.
For example, I'm seeing max Output energy 6548.4 for last night when the actual was around -5.2.
Don't know if this is related to guarding against -ve values as you mentioned above?

**Comment by TonyM1958** [comment]
OK. Just one other minor note. I looked at the price estimates and set this for Flux. The price boxes are in £ but the Flux prices are in p and a typical price is 16.51p/kWh. When you enter this as £0.1651 in the boxes, they overflow the line. Wondered if there might be a way to format or present the price that is a bit neater / more intuitive?

**Comment by alpriest** [comment]
Most users are now in the Eurozone so i've got to keep the price entry in € or £. I've moved that symbol outside the box though so it now accepts 4dp precision. This will be in 1.92 when that's released.
